### ISSUE TITLE
Invalid ArgumentError when training

### ISSUE THREAD

**Issue by huilinyang** [fkong7/MeshDeformNet#5]
Hello,

When I tried to train the model with pre-processed data, I ran into this error:

![image](https://user-images.githubusercontent.com/20769781/195659406-d6e31fa5-9d1d-4059-a307-b93983097d1a.png)

Could you help me with that? 

Your help will be greatly appreciated!

**Comment by huilinyang** [comment]
This is the same type of error I ran into when I tried twice.

![image](https://user-images.githubusercontent.com/20769781/195661023-155775a7-48cd-4d73-a64b-56da6bc5bebd.png)

**Comment by fkong7** [comment]
Could you provide the full exception trace? Line 190 is not the cause here. Also, it seems to be an index out-of-bound issue -  are you training on GPU?

**Comment by huilinyang** [comment]
Yes, I am trying to train on GPU. The full exception trace is as below.
![image](https://user-images.githubusercontent.com/20769781/195977456-f46cb438-b940-43a7-8d6e-6d12142896ae.png)

**Comment by huilinyang** [comment]
And this

![image](https://user-images.githubusercontent.com/20769781/195982675-05ef91a6-1f5e-4790-910f-103d91c551d3.png)

**Comment by huilinyang** [comment]
It seems the problem lies here:
The code in the laplace_coord(pred, feed_dict, block_id) function in loss.py. 

![image](https://user-images.githubusercontent.com/20769781/196027771-8a1e4cf3-4c9b-4239-90b4-66a7e70eec46.png)

The No.50 line comments that 'add one zero vertex since the laplace index was initialized to be -1'. The min of the lape_idx is -1 and the max is 11493. 

Can I simply add 1 to every element in lape_idx? 

The following are some values of lape_idx and its range.
![image](https://user-images.githubusercontent.com/20769781/196028425-5b3c823d-764a-4d97-b9e2-38dd16434d74.png)

![image](https://user-images.githubusercontent.com/20769781/196028469-08522abd-b0a4-42fb-999e-6d4686f64b8e.png)

**Comment by huilinyang** [comment]
The problem should be tf.gather() in the No. 151 line with the red circle. But I am still working on fixing it. Could you give me some hints? Thanks a lot!!!!!

![image](https://user-images.githubusercontent.com/20769781/197395984-1f92f05a-72cd-4aba-930a-5ee75e80f84f.png)

**Comment by fkong7** [comment]
Could you let me know why you think the issue is in line 151? Also, I haven't encountered this error before, so an example with your input data that could reproduce this error would help. Are you using TF1?

**Comment by huilinyang** [comment]
Yes, my environment is with TensorFlow 1.15.0.

Because the error has occurred on node loss/lambda_18_loss/GatherV2_7, loss/lambda_loss/GatherV2_7, loss/lambda_3_loss/GatherV2_7 and loss/lambda_9_loss/GatherV2_7 ... All of these belong to the losses of the first deformation of the graph network and they all occurred at the node GatherV2_7, which is corresponding to the tf.gather( ) in line 151. 

Cause I am a new man for Tensorflow, do you know how to debug while training the model? Specifically, how to check the value for each node when training? Thanks!!!

**Comment by huilinyang** [comment]
Sorry. I made a mistake. The GatherV2_7 should correspond to line 55 of loss.py as shown below.
It is because the Laplace index was initialized to be -1. But for TensorFlow 1.15.0, the indices should be equal to or greater than 0. Can I simply add 1 to the Laplace index? Or could you briefly explain the logic of the index initialization? For example, whether the index [-1] points to the first vertex, the index [0] points to the second vertex, and the index [1] points to the third vertex? etc. Or are they arranged by another logic? 

Thank you!!!

![image](https://user-images.githubusercontent.com/20769781/197760320-10a90b82-0bd9-4db5-bf85-10a3883251ca.png)

**Comment by fkong7** [comment]
I did some testing in TF 1.15.0, and using tf.gather with negative indices indeed caused errors at some occasions. Yes, I think you can add 1 to the indices, but remember to modify line 51 to concatenate zero vector **before** the predicted coordinates. The -1s  are currently used to slice the 0 vector at the end of `vertex` so that coordinates from non-neighboring vertices can be ignored.

**Comment by fkong7** [comment]
> Yes, my environment is with TensorFlow 1.15.0.
> 
> Because the error has occurred on node loss/lambda_18_loss/GatherV2_7, loss/lambda_loss/GatherV2_7, loss/lambda_3_loss/GatherV2_7 and loss/lambda_9_loss/GatherV2_7 ... All of these belong to the losses of the first deformation of the graph network and they all occurred at the node GatherV2_7, which is corresponding to the tf.gather( ) in line 151.
> 
> Cause I am a new man for Tensorflow, do you know how to debug while training the model? Specifically, how to check the value for each node when training? Thanks!!!

I think you could use tf.print to print out tensors during training.

**Comment by huilinyang** [comment]
> I did some testing in TF 1.15.0, and using tf.gather with negative indices indeed caused errors at some occasions. Yes, I think you can add 1 to the indices, but remember to modify line 51 to concatenate zero vector **before** the predicted coordinates. The -1s are currently used to slice the 0 vector at the end of `vertex` so that coordinates from non-neighboring vertices can be ignored.

Thank you so much. Your reply really helps a lot!
